python/ray/_private/runtime_env/conda_utils.py

```diff
@@ -128,42 +132,58 @@ class ShellCommandException(Exception):
 
 def exec_cmd(cmd: List[str],
              throw_on_error: bool = True,
-             stream_output: bool = False) -> Union[int, Tuple[int, str, str]]:
+             logger: Optional[logging.Logger] = None
+             ) -> Union[int, Tuple[int, str, str]]:
     """
     Runs a command as a child process.
 
     A convenience wrapper for running a command from a Python script.
 
-    Note on the return value: If stream_output is true, then only the exit code
-    is returned. If stream_output is false, then a tuple of the exit code,
+    Note on the return value: A tuple of the exit code,
     standard output and standard error is returned.
 
     Args:
         cmd: the command to run, as a list of strings
         throw_on_error: if true, raises an Exception if the exit code of the
             program is nonzero
-        stream_output: if true, does not capture standard output and error; if
-            false, captures these, streams and returns them
     """
-    if stream_output:
-        child = subprocess.Popen(
-            cmd, universal_newlines=True, stdin=subprocess.PIPE)
-        child.communicate()
-        exit_code = child.wait()
-        if throw_on_error and exit_code != 0:
-            raise ShellCommandException("Non-zero exitcode: %s" % (exit_code))
-        return exit_code
-    else:
-        child = subprocess.Popen(
-            cmd,
-            stdout=subprocess.PIPE,
-            stdin=subprocess.PIPE,
-            stderr=subprocess.PIPE,
-            universal_newlines=True)
-        (stdout, stderr) = child.communicate()
-        exit_code = child.wait()
-        if throw_on_error and exit_code != 0:
-            raise ShellCommandException(
-                "Non-zero exit code: %s\n\nSTDOUT:\n%s\n\nSTDERR:%s" %
-                (exit_code, stdout, stderr))
-        return exit_code, stdout, stderr
+    child = subprocess.Popen(
+        cmd,
+        stdout=subprocess.PIPE,
+        stdin=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        universal_newlines=True)
+    (stdout, stderr) = child.communicate()
+    exit_code = child.wait()
+    if throw_on_error and exit_code != 0:
+        raise ShellCommandException(
+            "Non-zero exit code: %s\n\nSTDOUT:\n%s\n\nSTDERR:%s" %
+            (exit_code, stdout, stderr))
+    return exit_code, stdout, stderr
+
+
+def exec_cmd_stream_to_logger(cmd: List[str],
+                              logger: logging.Logger,
+                              n_lines: int = 10) -> Tuple[int, str]:
```

Comment:
This is great!  In case you tested this manually, did you happen to see if 10 lines is enough to capture the relevant pip message if a bad pip package is provided?

Reply:
Ah, looks like this is covered in the unit test

Reply:
Yep it's already covered by the test you added!